kpatch-build/create-klprela-module.c

```diff
@@ -0,0 +1,400 @@
+	klpsec->data->d_type = ELF_T_RELA;
+	/* sh_info and sh_link are sent when rebuilding relasecs */
+
+	ALLOC_LINK(kr, &klp_relasecs);
+	kr->sec = klpsec;
+	kr->base = base;
+	kr->objname = objname;
+
+	return klpsec;
+}
+
+static void kpatch_unpack_krelas_ksyms(struct kpatch_elf *kelf, struct section *krelasec,
+				       struct section *ksymsec, char *strings)
+{
+	struct section *base, *klp_relasec;
+	struct kpatch_relocation *krelas;
+	struct symbol *sym;
+	struct rela *rela;
+	char *objname;
+	int index, nr;
+
+	krelas = krelasec->data->d_buf;
+	nr = krelasec->data->d_size / sizeof(*krelas);
+
+	for (index = 0; index < nr; index++) {
+		list_for_each_entry(rela, &krelasec->rela->relas, list) {
+			switch(rela->offset - (index * sizeof(*krelas))) {
+			case offsetof(struct kpatch_relocation, dest):
+				base = rela->sym->sec;
+				if (!base)
+					ERROR("base sec of krela not found");
+				break;
+
+			case offsetof(struct kpatch_relocation, objname):
+				objname = strdup(strings + rela->addend);
+				if (!objname)
+					ERROR("strdup");
+				break;
+
+			case offsetof(struct kpatch_relocation, ksym):
+				sym = kpatch_add_ksym_to_symbols(kelf, ksymsec,
+								 strings, rela->addend);
+				if (!sym)
+					ERROR("error adding ksym to symtab");
+				break;
+			}
+		}
+
+		if (!base || !objname || !sym)
+			ERROR("incomplete krela entry?");
+
+		klp_relasec = kpatch_find_klp_relasec(kelf, base, objname);
+		if (!klp_relasec)
+			ERROR("error finding klp relasec");
+
+		ALLOC_LINK(rela, &klp_relasec->relas);
+		rela->sym = sym;
+		rela->type = krelas[index].type;
+		rela->offset = krelas[index].offset;
+		rela->addend = krelas[index].addend;
+	}
+}
+
+static void kpatch_create_klp_arch_sections(struct kpatch_elf *kelf, char *strings)
+{
+	struct section *karch, *base;
+	struct kpatch_arch *entries;
+	struct rela *rela;
+	char *objname, *newname, *relaname;
+	int nr, len, index;
+
+	karch = find_section_by_name(&kelf->sections, ".kpatch.arch");
+	if (!karch)
+		return;
+
+	entries = karch->data->d_buf;
+	nr = karch->data->d_size / sizeof(*entries);
+
+	for (index = 0; index < nr; index++) {
+		list_for_each_entry(rela, &karch->rela->relas, list) {
+			switch(rela->offset - (index * sizeof(*entries))) {
+			case offsetof(struct kpatch_arch, sec):
+				base = rela->sym->sec;
+				if (!base)
+					ERROR("base sec of karch entry not found");
+				break;
+
+			case offsetof(struct kpatch_arch, objname):
+				objname = strdup(strings + rela->addend);
+				if (!objname)
+					ERROR("strdup");
+				break;
+			}
+		}
+
+		if (!base || !objname)
+			ERROR("no base or objname found for kpatch_arch entry");
+
+		len = strlen(KLP_ARCH_PREFIX) + strlen(objname) + \
+		      strlen(base->name) + 2;
+		newname = malloc(len);
+		if (!newname)
+			ERROR("malloc");
+		strcpy(newname, KLP_ARCH_PREFIX);
+		strcat(newname, objname);
+		strcat(newname, ".");
+		strcat(newname, base->name);
+		base->name = newname;
```

Comment:
Yeah, that sounds good.  Or maybe the opposite, make .klp.arch creation the default and have a --no-klp-arch-sections flag, to make it more futuristic.